### DISCUSSION TITLE
Persistence of UDFs

### DISCUSSION THREAD

**Discussion by Admolly** [duckdb/duckdb#8684] [General]
Right now, it is not possible to persist any UDF outside of the session during which it is created.  In the future, will this be possible?  I know there are plans for extending what has already been created, but this would be super helpful, especially when sharing a database file with another.

**Comment by Mause** [comment]
It could be possible in theory in interpreted languages, or where the source is available, such as JavaScript, nodejs, or Python, but it would require that you open the database using the same binding you originally created it with

**Reply by Admolly** [reply]
OK, so not likely then ;)

**Comment by stevenpelley** [comment]
I'm running into similar challenges when sharing files, but also when using multiple applications to connect to duckdb.  For example, when using data visualization tools such as apache superset for data exploration there's no easy way I've found to inject UDFs, which would require modifying superset.  This makes the development cycle tedious as I have to use other tools to query and materialize the results of my UDFs, and then go back to visualization and data exploration.  Persisted (or maybe "database-hosted") UDFs would solve this.

This is my "+1": letting you know that someone else out there is looking for this feature.

In addition to providing the source code and using the same binding I could see this being accomplished by:

- including a language runtime in duckdb, likely in an extension.  This makes the most sense for languages catered for embedding (javascript, web assembly, lua).  The source code or compiled webassembly would be stored in the database file.
- allowing duckdb to communicate with an external process _in the middle of processing SQL_.  Duckdb quacks arrow, but as far as I can tell only at the boundary of SQL queries, and still only within the same process.  (I'm not saying this is easy) Arrow data could be streamed via files, named pipes, mmap shared memory back and forth between the query engine and an external process with the semantics of a UDF.  I'd prioritize seamless over performant.  The UDF object would store information needed for the integration such as named pipes or filenames for shared arrow data, or these could be mandatory arguments to such a UDF (or controlled via .once-line pragmas).

**Reply by Mause** [reply]
You might have some luck with something like @MarkRoddy's  https://github.com/MarkRoddy/duckdb-pytables ?

**Reply by MarkRoddy** [reply]
Thanks for the shout out @Mause !

@stevenpelley PyTables is moving in this direction. It doesn't yet let you package the UDF into a database file, but that is on the roadmap. The extension does let you using Python functions as UDF's (undocumented atm) and as Table Functions. You'd have to copy said python code (and any dependencies) along with your database file for the time being. Happy to chat if that sounds useful to you.
